iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl

```diff
@@ -169,11 +169,11 @@ optional<T>::operator=(U&& newValue) noexcept
 {
     if (m_hasValue)
     {
-        value() = std::forward<T>(newValue);
```

Comment:
I think we have to revert this change. Maybe the solution is to add explicit "copy/move" assignment operators for the underlying value like:

```cxx
// in inl file
template <typename T>
inline optional<T>& optional<T>::operator=(const T& newValue) noexcept
{
    if (m_hasValue)
    {
        value() = newValue;
    }
    else
    {
        construct_value(newValue);
    }
    return *this;
}

template <typename T>
inline optional<T>& optional<T>::operator=(T&& newValue) noexcept
{
    if (m_hasValue)
    {
        value() = std::move(newValue);
    }
    else
    {
        construct_value(std::move(newValue));
    }
    return *this;
}
```

and add the definition to the hpp file:
```cxx
optional& operator=(const T& newValue) noexcept;
optional& operator=(T&& newValue) noexcept;
```

Reply:
Taking a look at https://github.com/eclipse-iceoryx/iceoryx/blob/1e17b23d50e7ac2243f43969d90e7f5e8b60afca/tools/introspection/source/introspection_app.cpp#L796 where one of the errors comes from, it seems like a side effect of the current behavior is to implicitly move when assigning a non-const reference.

The `std::forward<T>` turns into `T&&`, while `std::forward<U>` with `U = T&` is equivalent to `T&`.  This matches the copy constructor of `Sample` (which is deleted) rather than the move constructor.  It seems like this is not desired behavior (implicitly converting an l-value to r-value reference).  I can fix it here by explicitly doing `memPoolSample = std::move(sample)`, though I'm not sure what implications this might have elsewhere in Iceoryx or for other users.

Reply:
This was the only file I found with issues when building locally, so I pushed an update to try again.  Let me know how you'd like to proceed if CI passes.

Reply:
We should realize this quickly when this was the wrong approach, so I would go for it and test the introspection app afterwards. If it still shows all the mempools we should be fine.

In my opinion, we are with iceoryx at a state with testing and the CI that we can perform fearless refactoring and rely on the CI.

Reply:
I ran the introspection client with the changes, along with just `iox-roudi` and I see this output: 
![image](https://github.com/eclipse-iceoryx/iceoryx/assets/156106487/b6a81d34-9600-4c5c-958b-e9910a0fe363)

Reply:
Mempool looks exactly like it should. Nice.